galley/pkg/config/analysis/analyzers/auth/mtls.go

```diff
@@ -0,0 +1,333 @@
+var _ analysis.Analyzer = &MTLSAnalyzer{}
+
+// Metadata implements Analyzer
+func (s *MTLSAnalyzer) Metadata() analysis.Metadata {
+	return analysis.Metadata{
+		Name: "auth.MTLSAnalyzer",
+
+		// Each analyzer should register the collections that it needs to use as input.
+		Inputs: collection.Names{
+			metadata.K8SCoreV1Pods,
+			metadata.K8SCoreV1Namespaces,
+			metadata.K8SCoreV1Services,
+			metadata.IstioAuthenticationV1Alpha1Meshpolicies,
+			metadata.IstioAuthenticationV1Alpha1Policies,
+			metadata.IstioNetworkingV1Alpha3Destinationrules,
+		},
+	}
+}
+
+// Analyze implements Analyzer
+func (s *MTLSAnalyzer) Analyze(c analysis.Context) {
+	// TODO Reuse pilot logic as a library rather than reproducing its logic
+	// here.
+
+	// If autoMTLS is turned on, bail out early as the logic used below does not
+	// reason about its usage.
+	autoMtlsEnabled := false
+	c.ForEach(metadata.IstioMeshV1Alpha1MeshConfig, func(r *resource.Entry) bool {
+		mc := r.Item.(*meshconfig.MeshConfig)
+		if mc.GetEnableAutoMtls() != nil && mc.GetEnableAutoMtls().Value {
+			autoMtlsEnabled = true
+		}
+		return true
+	})
+	if autoMtlsEnabled {
+		return
+	}
+
+	// Loop over all services, building up a list of selectors for each. This is
+	// used to determine which pods are in which services, and determine whether
+	// or not the sidecar is fully enmeshed. If a service doesn't have a
+	// sidecar, then we always treat it as having an explicit "plaintext" policy
+	// regardless of the service/namespace/mesh policy.
+
+	var targetServices []mtls.TargetService
+	fqdnsWithoutSidecars := make(map[string]struct{})
+	// Keep track of each fqdn -> port name -> port number. This is because
+	// the Policy object lets you target a port name, but DR requires port
+	// number. Tracking this means we can normalize to port number later.
+	fqdnToNameToPort := make(map[string]map[string]uint32)
+
+	// TODO store portname!
+	c.ForEach(metadata.K8SCoreV1Services, func(r *resource.Entry) bool {
+		svcNs, svcName := r.Metadata.Name.InterpretAsNamespaceAndName()
+		svc := r.Item.(*v1.ServiceSpec)
+
+		svcSelector := k8s_labels.SelectorFromSet(svc.Selector)
+		fqdn := util.ConvertHostToFQDN(svcNs, svcName)
+		for _, port := range svc.Ports {
+			portNumber := uint32(port.Port)
+			// portName is optional, but we note it so we can translate later.
+			if port.Name != "" {
+				// allocate a new map if necessary
+				if _, ok := fqdnToNameToPort[fqdn]; !ok {
+					fqdnToNameToPort[fqdn] = make(map[string]uint32)
+				}
+				fqdnToNameToPort[fqdn][port.Name] = portNumber
+			}
+
+			targetServices = append(targetServices, mtls.NewTargetServiceWithPortNumber(fqdn, portNumber))
+		}
+
+		// Now we loop over all pods looking for sidecars that match our
+		// service. If we find a single pod without a sidecar, we label the
+		// service as not having a sidecar (which means we bypass policy
+		// checking).
+		c.ForEach(metadata.K8SCoreV1Pods, func(pr *resource.Entry) bool {
```

Comment:
> This implicitly means that this analyzer will basically never work in the files-only case, right? Since pods are not generally directly defined in files.

You are right - it is not very useful when run against files alone (although the global mesh config is still checked, so if you messed up your global policy that can still be detected early).

It would be cool to be able to reason about how Deployments, Jobs, ReplicaSets etc all create workloads that would/wouldn't have a sidecar and then work in the file-only case, but it would be pretty complex as well :/

> BTW I appreciate your thorough code comments and careful decomposition of the various layers at work. You've made this really hairy logic that we're trying to match a lot more approachable than it would be otherwise!

Thank you!